### ISSUE TITLE
Add support for psycopg2 driver.

### ISSUE THREAD

**Issue by jackwotherspoon** [GoogleCloudPlatform/cloud-sql-python-connector#214]
Support PostgreSQL [psycopg2](https://github.com/psycopg/psycopg2/) driver with connector.

**Comment by gunar** [comment]
Hey @jackwotherspoon, I've hit this exact problem this week. Do you have an estimate? Any rough ideas would be greatly appreciated. Thanks in advance.

**Comment by jackwotherspoon** [comment]
@gunar This has been prioritized and will be worked on in early 2022. 

In the meantime, you can connect to Cloud SQL using Python and psycopg2 with the Cloud SQL Auth Proxy through Unix sockets or TCP. For an explanation and example see here: https://cloud.google.com/sql/docs/postgres/connect-admin-proxy#expandable-2

For the Unix sockets sample just swap out part of the pg8000 code for this code snippet:
```
pool = sqlalchemy.create_engine(
    sqlalchemy.engine.url.URL.create(
        drivername="postgresql+psycopg2",
        username=db_user,  # e.g. "my-database-user"
        password=db_pass,  # e.g. "my-database-password"
        database=db_name,  # e.g. "my-database-name"
        query={
            "host: "{}/{}".format(
                db_socket_dir,  # e.g. "/cloudsql"
                instance_connection_name)  # i.e "<PROJECT-NAME>:<INSTANCE-REGION>:<INSTANCE-NAME>"
        }
    ),
    **db_config
)
```

Hope this helps for the time being! Thanks for the feedback!

**Comment by gunar** [comment]
@jackwotherspoon Thanks man. I'm not inclined to go with the Cloud SQL Auth Proxy as that's one more thing for the user to remember to execute.

I guess I could make the python script itself spawn the process for the proxy—has anyone attempted that?

Also, do I need `sqlalchemy` in order to connect to the proxy or can I connect directly with `psycopg2`?

**Comment by kemics** [comment]
Hi @jackwotherspoon , thank you for all your work on cloud sql connectors

Do you have any estimates for psycopg2 support? It would be really nice to have

By the way the issue has p2 priority but it has the biggest number of thumb-ups😉

**Comment by jackwotherspoon** [comment]
Hi @kemics, thanks for the kind words and showing your enthusiasm towards the Cloud SQL Python connector.

I currently can't provide a specific timeline for psycopg2 support because there are a lot of factors at play here but I can provide some details for you. 

In order to support psycopg2 connections with the connector we require the ability to pass in a pre-existing SSL connection and skip a couple of postgres database level protocols (this is the hard part and is required because Cloud SQL connections connect to a proxy server and not directly to the database itself).

I have been in talks with the psycopg2 team about helping make some changes in psycopg2 to help support these type of connections, the work also requires additional upstream changes into [libpq](https://www.postgresql.org/docs/current/libpq.html).

Overall, there are a lot of moving parts but I will continue to keep the ball rolling on this and make sure that I update progress here on the issue for more transparency.

Thanks again for the gentle nudge here and for showing interest in the Python connector, much appreciated.

Have a great day :)

**Comment by kemics** [comment]
Thanks @jackwotherspoon , great explanation 👍

I'm a fan of what you are doing here and have successfully tried this repo with sqlalchemy and also https://github.com/GoogleCloudPlatform/cloud-sql-go-connector . Works like a charm and a huge improvement in terms of security

**Comment by gunar** [comment]
@jackwotherspoon 

With regards to:

> For the Unix sockets sample just swap out part of the pg8000 code for this code snippet:

Is there any way to use my already logged in `gcloud` IAM authentication and not have to pass in username and password as part of the code? (I'm trying to get away from `.env` files).

**Comment by jackwotherspoon** [comment]
@gunar
> Is there any way to use my already logged in `gcloud` IAM authentication and not have to pass in username and password as part of the code? (I'm trying to get away from `.env` files).

This can be done in a fairly straight-forward way for a service account authentication using [Manual IAM Database Authentication](https://cloud.google.com/sql/docs/postgres/authentication). You just need to [add your service account as an IAM database user](https://cloud.google.com/sql/docs/postgres/add-manage-iam-users#creating-a-database-user) to your Cloud SQL instance and grant it the **Cloud SQL Instance User** IAM role alongside the usual **Cloud SQL Client** role.

Once this is done, the below snippet should work in any environment that uses the service account (i.e. Cloud Run, Cloud Functions, App Engine etc.). See comments below for running locally.

```python
import google.auth
from google.auth.transport.requests import Request
import sqlalchemy
import psycopg2

# uncomment below two lines to set application credentials to service account locally
# import os
# os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "path/to/key.json"

# get application default credentials
creds, _ = google.auth.default()

# refresh credentials if not valid
if not creds.valid:
  request = Request()
  creds.refresh(request)

# IAM service account database user, postgres removes suffix
sa_iam_user = (creds.service_account_email).removesuffix(
  ".gserviceaccount.com"
)

# create SQLAlchemy connection pool
pool = sqlalchemy.create_engine(
  sqlalchemy.engine.url.URL.create(
    drivername="postgresql+psycopg2",
    username=sa_iam_user, # login to database as service account user
    password=str(creds.token), # manual IAM database authn
    database=db_name,  # e.g. "my-database-name"
    query={
      "host": "{}/{}".format(
          "/cloudsql",  # e.g. "/cloudsql"
          instance_connection_name)  # i.e "<PROJECT-NAME>:<INSTANCE-REGION>:<INSTANCE-NAME>"
    }
  ),
)

# ... use connection pool
```

**Comment by cw-optima-ovo** [comment]
I've been struggling to get psycopg2 working with either of the solutions posted here, really appreciate the examples @jackwotherspoon - wondering if you or someone else could point me in the right direction...
I need to connect to a Cloud SQL instance running Postgres with a gen2 cloud function using an IAM service account. Tried the solution posted on 16/12/21 as well as the solution posted 3/8/22 but still not getting there.

The cloud function is connected to the VPC with a serverless VPC connector, and the cloud sql instance has a private and public IP. I've also tried the below with the private IP instead of the 'project-name:instance-region:instance-name' for the unix socket.

I'm getting the error
```psycopg2.OperationalError: connection to server on socket "/cloudsql/project-name:instance-region:instance-name/.s.PGSQL.5432" failed: No such file or directory```


This is my code (obfuscated of course)

```
    # get application default credentials
    creds, _ = google.auth.default()

    # refresh credentials if not valid
    if not creds.valid:
        request = Request()
        creds.refresh(request)

    # IAM service account database user, postgres removes suffix
    sa_iam_user = 'service-account@project-name.iam'

    # create SQLAlchemy connection pool
    engine = sqlalchemy.create_engine(
    sqlalchemy.engine.url.URL.create(
        drivername="postgresql+psycopg2",
        username=sa_iam_user, # login to database as service account user
        password=str(creds.token), # manual IAM database authn
        database='database-name',  # e.g. "my-database-name"
        query={
        "host": "{}/{}".format(
            "/cloudsql",  # e.g. "/cloudsql"
            # 'project-name:instance-region:instance-name')  # i.e "<PROJECT-NAME>:<INSTANCE-REGION>:<INSTANCE-NAME>"
        }
    ),
    )
```

What am I doing wrong? I feel like I'm very close but obviously missing something.

**Comment by jackwotherspoon** [comment]
@cw-optima-ovo Let me see if I can help you out! Thanks for raising your issue and sorry for the slow reply.

I should of made one thing clear in my previous comment with the code sample. The Unix sockets approach is for a **Public IP** connection. 

The recommended approach for **Private IP** connections such as your own is through a TCP socket which I will showcase here: (code sample was tested for gen2 Cloud Functions with VPC Connector attached)

```python
import google.auth
from google.auth.transport.requests import Request
import sqlalchemy
import psycopg2

# get application default credentials
creds, _ = google.auth.default()

# refresh credentials if not valid
if not creds.valid:
  request = Request()
  creds.refresh(request)

# IAM service account database user, postgres removes suffix
sa_iam_user = (creds.service_account_email).removesuffix(
  ".gserviceaccount.com"
)

# create SQLAlchemy connection pool
pool = sqlalchemy.create_engine(
    # Equivalent URL:
    # postgresql+psycopg2://<db_user>:<db_pass>@<db_host>:<db_port>/<db_name>
    sqlalchemy.engine.url.URL.create(
        drivername="postgresql+psycopg2",
        username=sa_iam_user,
        password=str(creds.token),
        host="YOUR_PRIVATE_IP_ADDRESS",
        port=5432,
        database="YOUR_DATABASE",
    ),
)
# ... use connection pool
```

Let me know if you have luck with this sample over the unix sockets version 😄  Make sure the service account that the Cloud Function uses is added as a proper Cloud SQL IAM user to the instance and has the proper roles.

**Comment by cw-optima-ovo** [comment]
> @cw-optima-ovo Let me see if I can help you out! Thanks for raising your issue and sorry for the slow reply.
> 
> I should of made one thing clear in my previous comment with the code sample. The Unix sockets approach is for a **Public IP** connection.
> 
> The recommended approach for **Private IP** connections such as your own is through a TCP socket which I will showcase here: (code sample was tested for gen2 Cloud Functions with VPC Connector attached)
> 
> ```python
> import google.auth
> from google.auth.transport.requests import Request
> import sqlalchemy
> import psycopg2
> 
> # get application default credentials
> creds, _ = google.auth.default()
> 
> # refresh credentials if not valid
> if not creds.valid:
>   request = Request()
>   creds.refresh(request)
> 
> # IAM service account database user, postgres removes suffix
> sa_iam_user = (creds.service_account_email).removesuffix(
>   ".gserviceaccount.com"
> )
> 
> # create SQLAlchemy connection pool
> pool = sqlalchemy.create_engine(
>     # Equivalent URL:
>     # postgresql+psycopg2://<db_user>:<db_pass>@<db_host>:<db_port>/<db_name>
>     sqlalchemy.engine.url.URL.create(
>         drivername="postgresql+psycopg2",
>         username=sa_iam_user,
>         password=str(creds.token),
>         host="YOUR_PRIVATE_IP_ADDRESS",
>         port=5432,
>         database="YOUR_DATABASE",
>     ),
> )
> # ... use connection pool
> ```
> 
> Let me know if you have luck with this sample over the unix sockets version 😄 Make sure the service account that the Cloud Function uses is added as a proper Cloud SQL IAM user to the instance and has the proper roles.

Thanks @jackwotherspoon that does indeed work! However I ended up going down the asyncpg route with the sql python connector in the end, the reason for this was partly because I didn't want to have a fixed IP in my variables as we're deploying a project through CI/CD. So while it's solvable to get the IP with CI/CD, it's a headache we didn't want to introduce given we had the project:region:instance already available. The performance between asyncpg and psycopg2 is about the same for our context so no harm done.

Hoping to see the psycopg2 supported with the cloud sql connector in the future though, this would have been my original preference.!

**Comment by jackwotherspoon** [comment]
> Can I add a request for psycopg3 here? Or should that go in a separate issue?

@wwuck  psycopg3 is already being tracked in a separate issue https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/issues/219, although both will most likely be supported at the same time once we can get up-stream changes to libpq made.

**Comment by jackwotherspoon** [comment]
FYI to everyone: it is possible to connect with automatic IAM authentication without the need for the Python Connector:

You'll need to ensure a few things:

- The token has only `sql.login` scope (i.e. `https://www.googleapis.com/auth/sqlservice.login`)
- The token isn't transmitted over an unencrypted channel. (Private IP recommended)

We're working on making this path easier for folks, but for now I'll share the mechanics for visibility.

Assuming you're using [SQLAlchemy](https://www.sqlalchemy.org/), you can do this:
```python
import sqlalchemy
from sqlalchemy import event

import google.auth
from google.auth.transport.requests import Request

# initialize ADC creds
creds, _ = google.auth.default(
    scopes=["https://www.googleapis.com/auth/sqlservice.login"]
)

# Cloud SQL Instance IP address
instance_ip = <INSTANCE_IP> 

# create SQLAlchemy connection pool
# use Cloud SQL Instance IP + native port (5432)
# for best security use client certificates + server cert for SSL
engine = sqlalchemy.create_engine(
    f"postgresql+psycopg2://service-account@project-id.iam:empty-pass@{ip_address}:5432/dbname",
    connect_args={'sslmode': 'require'},
)

# set do_connect event listener
@event.listens_for(engine, 'do_connect')
def auto_iam_authn(dialect, conn_rec, cargs, cparams):  
    # refresh credentials if expired
    if not creds.valid:
        request = Request()
        creds.refresh(request)
    
    # use oauth2 token as password
    cparams["password"] = str(creds.token)

# interact with Cloud SQL database using connection pool
with engine.connect() as conn:
    # query database
    time = conn.execute(sqlalchemy.text("SELECT NOW()")).fetchone()
    print(time[0])
```

**Comment by bapi24** [comment]
@jackwotherspoon is there a way connecting to the instance with IAM without using `sqlalchemy` but using just `psycopg2`?
```
# example
with psycopg2.connect(url) as con:
    with con.cursor() as cur:
        cur.execute("SELECT * FROM test.table")
        rows = cur.fetchall()
        print(rows)
```

**Comment by jackwotherspoon** [comment]
> @jackwotherspoon is there a way connecting to the instance with IAM without using `sqlalchemy` but using just `psycopg2`?

Yes, you can still use the same concept as above. You can set the `password` to that of the IAM Principal's OAuth2 token. You will just have to manage refreshing the credentials yourself within your application to make sure the credentials object has not expired.

However, we normally recommend taking advantage of connection pooling for production environment, thus we strongly recommend the above SQLAlchemy example for most cases.

```python
import psycopg2

import google.auth
from google.auth.transport.requests import Request

# initialize ADC creds
creds, _ = google.auth.default(
    scopes=["https://www.googleapis.com/auth/sqlservice.login"]
)

 # refresh credentials if expired (manage this code in your application)
if not creds.valid:
    request = Request()
    creds.refresh(request)


# Cloud SQL Instance IP address
instance_ip = '<INSTANCE_IP>'

# interact with Cloud SQL database using psycopg2 connection
with psycopg2.connect(f"dbname=mydb user=sa-name@project-id.iam password={str(creds.token)} host={instance_ip} sslmode=require") as con:
    with con.cursor() as cur:
        cur.execute("SELECT * FROM test.table")
        rows = cur.fetchall()
        print(rows)
```

**Comment by jackwotherspoon** [comment]
@bapi24 Glad it worked! Let me answer your questions 😄 

>  The documentation below says to use `cloudsql.instances.login` IAM permission to a custom role. If we attach that to a service-account, would service-account be able to talk to the DB same way?

The above code can be used for both an IAM User or IAM service account! So yes, you can add your service account as an IAM database user and have it login and access your DB 😄 

> Regarding the way we are getting the token here, is this a workaround for doing `gcloud auth application-default login`?

No it is not a workaround as the two perform slightly different functions (that work together). The way we are getting the token in code is *fetching* the Application Default Credentials from the environment and adding the required scopes for the OAuth2 token to be able to login to the DB. The `gcloud auth application-default login` command is *setting* the Application Default Credentials for your local environment. This is the key distinction, one *gets* the ADC credentials and one *sets* the ADC credentials.

The above code will work locally and in an application hosted in the Cloud (Cloud Run, App Engine etc.) since the ADC credentials in the Cloud are set automatically for you to the service account used with the application (no need to run a gcloud auth command).

> would `psycopg` be supported in [`cloudsql-python-connector`](https://github.com/GoogleCloudPlatform/cloud-sql-python-connector) anytime soon?

It is most likely that psycopg will not be supported anytime soon in the Cloud SQL Python Connector (if ever) due to the vast overhead required to patch the psycopg library as well as the libpq library. This work coupled with the fact that the above samples (SQLAlchemy and direct psycopg2) already work very nicely without the need for the Python Connector makes this a low priority effort.

**Comment by bapi24** [comment]
Got it, thank you very much for explaining it in detail 🙌

**Comment by AndreasBergmeier6176** [comment]
> FYI to everyone: it is possible to connect with automatic IAM authentication without the need for the Python Connector:
> 
> You'll need to ensure a few things:
> 
>     * The token has only `sql.login` scope (i.e. `https://www.googleapis.com/auth/sqlservice.login`)
> 
>     * The token isn't transmitted over an unencrypted channel. (Private IP recommended)
> 
> 
> We're working on making this path easier for folks, but for now I'll share the mechanics for visibility.
> 
> Assuming you're using [SQLAlchemy](https://www.sqlalchemy.org/), you can do this:
> 
> ```python
> import sqlalchemy
> from sqlalchemy import event
> 
> import google.auth
> from google.auth.transport.requests import Request
> 
> # initialize ADC creds
> creds, _ = google.auth.default(
>     scopes=["https://www.googleapis.com/auth/sqlservice.login"]
> )
> 
> # Cloud SQL Instance IP address
> instance_ip = <INSTANCE_IP> 
> 
> # create SQLAlchemy connection pool
> # use Cloud SQL Instance IP + native port (5432)
> # for best security use client certificates + server cert for SSL
> engine = sqlalchemy.create_engine(
>     f"postgresql+psycopg2://service-account@project-id.iam:empty-pass@{ip_address}:5432/dbname",
>     connect_args={'sslmode': 'require'},
> )
> 
> # set do_connect event listener
> # This gets called before a connection is created and allows you to
> # refresh the OAuth2 token here as needed
> @event.listens_for(engine, 'do_connect')
> def auto_iam_authn(dialect, conn_rec, cargs, cparams):  
>     # refresh credentials if expired
>     if not creds.valid:
>         request = Request()
>         creds.refresh(request)
>     
>     # use OAuth2 token as password
>     cparams["password"] = str(creds.token)
> 
> # interact with Cloud SQL database using connection pool
> with engine.connect() as conn:
>     # query database
>     time = conn.execute(sqlalchemy.text("SELECT NOW()")).fetchone()
>     print(time[0])
> ```

Thanks for providing these examples. Although it would be really good if they were in the official documentation.

One thing that we still cannot make work is doing the psycopg2 -> Google SQL Auth Proxy -> Postgres connection in GKE using automatic IAM. I was under the impression that you would not need the login token there - am I wrong?

**Comment by enocom** [comment]
Correct -- the Proxy will handle that for you. If you'd like to open an issue in the [Proxy repo](https://github.com/GoogleCloudPlatform/cloud-sql-proxy), I'd be happy to help you out.

The approach describe above assumes you bypass the Connector and Proxy entirely.

**Comment by nordewal** [comment]
We are currently using the Auth Proxy binary in a separate sidecar container, but would love to only use this library for simplicity. Are there still plans to add support for the psycopg2 driver?

**Comment by jackwotherspoon** [comment]
> We are currently using the Auth Proxy binary in a separate sidecar container, but would love to only use this library for simplicity. Are there still plans to add support for the psycopg2 driver?

@nordewal Yes it is still in our backlog and on our radar! Unfortunately, I have just have not had the time to tackle it.

That being said, if any external contributors want to take a first attempt at it we would happily take any contributions 😄 

Happy to assist on providing guidance on the approach I would take.

**Comment by jeffclay** [comment]
FYI, it seems that psycopg2 is no longer getting new features and was superseded by pyscopg version 3. https://github.com/psycopg/psycopg2/commit/89005ac5b849c6428c05660b23c5a266c96e677d

**Comment by areiner222** [comment]
Any update on this?

**Comment by UzielSilva** [comment]
@jackwotherspoon I would like to contribute on working on this. What should I do to get some more context about?

**Comment by enocom** [comment]
This might not be a great first issue to work on, as psycopg has the same limitations as the underlying libpq library and so doesn't expose an API to create sockets for the connection to Cloud SQL.

High level, we've been thinking that to support psycopg we would create a unix socket within the Python connector,  and configure psycopg with that socket. Then internally, we would proxy data to and from the unix socket to the TLS connection to the Cloud SQL instance. We've done this in the Node connector, but it's a non-trivial change.

**Comment by UzielSilva** [comment]
I can give it a try, I have the files already created and yesterday was experimenting with psycopg constructors to see if they allow to pass a socket, the solutions that I have found so far are:
- Write temporary files having the TLS files and passing them to the parameters
- Open some sort of ssh tunneling, and make the connection happen through local host and ports
- The solution you mentioned
Going to give a shot to the solution you mentioned and I hope to get you back with a PR or with more questions 😄

**Comment by jeffclay** [comment]
@UzielSilva are you planning to implement legacy psycopg2 or psycopg version 3?

**Comment by UzielSilva** [comment]
I'm already working on psycopg 3

**Comment by UzielSilva** [comment]
@enocom I managed to create the unix proxy and the connector is now working with psycopg 3, the issue now is related to SQLAlchemy, it has some issues detecting the version and it also expects `show transaction isolation level` to return an array, which isn't. BUT, if I try to use the connection without SQLAlchemy, it works, actually it gives me back something like `{'transaction_isolation': 'read committed'}`.

I think this is huge progress, since psycopg 3 is already beaten, but I need some help making it to work with SQLAlchemist. I can setup a PR but of course tests are going to break. Do you want me to do a test that doesn't use SQLAlchemy?

**Comment by UzielSilva** [comment]
Adding my PR here: https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/pull/1311
